File: android-client/app/src/main/java/com/voicemic/app/Protocol.java
package com.voicemic.app;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;

/**
 * VoiceMic wire protocol — matches the Python server implementation.
 */
public class Protocol {

    public static final byte[] MAGIC = new byte[]{'V', 'M', 'I', 'C'};
    public static final int VERSION = 1;
    public static final int DEFAULT_PORT = 8125;
    public static final int HEADER_SIZE = 9; // MAGIC(4) + TYPE(1) + LENGTH(4)

    // Packet types
    public static final int PKT_HANDSHAKE = 0x01;
    public static final int PKT_HANDSHAKE_ACK = 0x02;
    public static final int PKT_AUDIO = 0x10;
    public static final int PKT_CONTROL = 0x20;
    public static final int PKT_PING = 0x30;
    public static final int PKT_PONG = 0x31;
    public static final int PKT_DISCONNECT = 0xFF;

    // Codec IDs
    public static final int CODEC_PCM = 0;
    public static final int CODEC_OPUS = 1;

    // Control commands
    public static final int CTRL_MUTE = 0x01;
    public static final int CTRL_UNMUTE = 0x02;
    public static final int CTRL_SET_VOLUME = 0x03;
    public static final int CTRL_SET_NOISE_SUPPRESS = 0x04;

    /**
     * Build a protocol packet: [MAGIC(4)][TYPE(1)][LENGTH(4)][PAYLOAD(N)]
     */
    public static byte[] buildPacket(int pktType, byte[] payload) {
        ByteBuffer buf = ByteBuffer.allocate(HEADER_SIZE + payload.length);
        buf.order(ByteOrder.BIG_ENDIAN);
        buf.put(MAGIC);
        buf.put((byte) pktType);
        buf.putInt(payload.length);
        buf.put(payload);
        return buf.array();
    }

    /**
     * Build handshake packet.
     */
    public static byte[] buildHandshake(int sampleRate, int channels, int codec, String deviceName) {
        byte[] nameBytes = deviceName.getBytes(StandardCharsets.UTF_8);
        int nameLen = Math.min(nameBytes.length, 64);

        ByteBuffer payload = ByteBuffer.allocate(8 + nameLen);
        payload.order(ByteOrder.BIG_ENDIAN);
        payload.putInt(sampleRate);
        payload.putShort((short) channels);
        payload.put((byte) codec);
        payload.put((byte) nameLen);
        payload.put(nameBytes, 0, nameLen);

        return buildPacket(PKT_HANDSHAKE, payload.array());
    }

    /**
     * Parse handshake ACK. Returns true if accepted.
     */
    public static boolean parseHandshakeAck(byte[] payload) {
        if (payload == null || payload.length < 1) return false;
        return payload[0] == 1;
    }

    /**
     * Build audio data packet.
     */
    public static byte[] buildAudioPacket(byte[] audioData) {
        return buildPacket(PKT_AUDIO, audioData);
    }

    /**
     * Build audio data packet from buffer segment.
     */
    public static byte[] buildAudioPacket(byte[] audioData, int offset, int length) {
        byte[] segment = new byte[length];
        System.arraycopy(audioData, offset, segment, 0, length);
        return buildPacket(PKT_AUDIO, segment);
    }

    /**
     * Build ping packet with current timestamp.
     */
    public static byte[] buildPing(long timestampMs) {
        ByteBuffer payload = ByteBuffer.allocate(8);
        payload.order(ByteOrder.BIG_ENDIAN);
        payload.putLong(timestampMs);
        return buildPacket(PKT_PING, payload.array());
    }

    /**
     * Parse pong packet, returns original timestamp.
     */
    public static long parsePong(byte[] payload) {
        if (payload == null || payload.length < 8) return 0;
        ByteBuffer buf = ByteBuffer.wrap(payload);
        buf.order(ByteOrder.BIG_ENDIAN);
        return buf.getLong();
    }

    /**
     * Build control packet.
     */
    public static byte[] buildControl(int command, int value) {
        byte[] payload = new byte[]{(byte) command, (byte) value};
        return buildPacket(PKT_CONTROL, payload);
    }

    /**
     * Build disconnect packet.
     */
    public static byte[] buildDisconnect() {
        return buildPacket(PKT_DISCONNECT, new byte[0]);
    }

    /**
     * Parse header from bytes. Returns [pktType, payloadLength] or null.
     */
    public static int[] parseHeader(byte[] header) {
        if (header.length < HEADER_SIZE) return null;

        // Validate magic
        for (int i = 0; i < 4; i++) {
            if (header[i] != MAGIC[i]) return null;
        }

        int pktType = header[4] & 0xFF;
        ByteBuffer buf = ByteBuffer.wrap(header, 5, 4);
        buf.order(ByteOrder.BIG_ENDIAN);
        int payloadLen = buf.getInt();

        return new int[]{pktType, payloadLen};
    }
}

File: pc-server/protocol.py
"""
VoiceMic Protocol Definition
Wire protocol for audio streaming between phone and PC.
"""
import struct

# Protocol constants
MAGIC = b"VMIC"
VERSION = 1
DEFAULT_PORT = 8125

# Packet types
PKT_HANDSHAKE = 0x01
PKT_HANDSHAKE_ACK = 0x02
PKT_AUDIO = 0x10
PKT_CONTROL = 0x20
PKT_PING = 0x30
PKT_PONG = 0x31
PKT_DISCONNECT = 0xFF

# Codec IDs
CODEC_PCM = 0
CODEC_OPUS = 1

# Control commands
CTRL_MUTE = 0x01
CTRL_UNMUTE = 0x02
CTRL_SET_VOLUME = 0x03
CTRL_SET_NOISE_SUPPRESS = 0x04

# Sample rates
SAMPLE_RATES = [8000, 16000, 22050, 44100, 48000]
DEFAULT_SAMPLE_RATE = 48000
DEFAULT_CHANNELS = 1
DEFAULT_FRAME_SIZE = 960  # 20ms at 48kHz


def build_packet(pkt_type: int, payload: bytes = b"") -> bytes:
    """Build a protocol packet: [MAGIC(4)][TYPE(1)][LENGTH(4)][PAYLOAD(N)]"""
    header = MAGIC + struct.pack("!BI", pkt_type, len(payload))
    return header + payload


def parse_packet_header(data: bytes):
    """Parse packet header, returns (pkt_type, payload_length) or None."""
    if len(data) < 9:
        return None
    magic = data[:4]
    if magic != MAGIC:
        return None
    pkt_type = data[4]
    payload_len = struct.unpack("!I", data[5:9])[0]
    return pkt_type, payload_len


def build_handshake(sample_rate: int, channels: int, codec: int, device_name: str = "Android") -> bytes:
    """Build handshake packet."""
    name_bytes = device_name.encode("utf-8")[:64]
    payload = struct.pack("!IHBB", sample_rate, channels, codec, len(name_bytes)) + name_bytes
    return build_packet(PKT_HANDSHAKE, payload)


def parse_handshake(payload: bytes):
    """Parse handshake payload. Returns dict with config."""
    if len(payload) < 8:
        return None
    sample_rate, channels, codec, name_len = struct.unpack("!IHBB", payload[:8])
    device_name = payload[8:8 + name_len].decode("utf-8", errors="replace")
    return {
        "sample_rate": sample_rate,
        "channels": channels,
        "codec": codec,
        "device_name": device_name,
    }


def build_handshake_ack(accepted: bool, message: str = "") -> bytes:
    """Build handshake acknowledgement."""
    msg_bytes = message.encode("utf-8")[:128]
    payload = struct.pack("!B", 1 if accepted else 0) + msg_bytes
    return build_packet(PKT_HANDSHAKE_ACK, payload)


def build_audio_packet(audio_data: bytes) -> bytes:
    """Build audio data packet."""
    return build_packet(PKT_AUDIO, audio_data)


def build_control(command: int, value: int = 0) -> bytes:
    """Build control packet."""
    payload = struct.pack("!BB", command, value)
    return build_packet(PKT_CONTROL, payload)


def build_ping(timestamp_ms: int) -> bytes:
    """Build ping packet."""
    payload = struct.pack("!Q", timestamp_ms)
    return build_packet(PKT_PING, payload)


def build_pong(timestamp_ms: int) -> bytes:
    """Build pong packet."""
    payload = struct.pack("!Q", timestamp_ms)
    return build_packet(PKT_PONG, payload)
